cmd/thanos/endpointset.go

```diff
@@ -0,0 +1,330 @@
+	content, err := configFile.Content()
+	if err != nil {
+		return EndpointConfig{}, errors.Wrapf(err, "unable to load config content: %s", configFile.Path())
+	}
+	var cfg EndpointConfig
+	if err := yaml.Unmarshal(content, &cfg); err != nil {
+		return EndpointConfig{}, errors.Wrapf(err, "unable to unmarshal config content: %s", configFile.Path())
+	}
+	for _, ecfg := range cfg.Endpoints {
+		if dns.IsDynamicNode(ecfg.Address) && ecfg.Strict {
+			return EndpointConfig{}, errors.Newf("%s is a dynamically specified endpoint i.e. it uses SD and that is not permitted under strict mode.", ecfg.Address)
+		}
+		if dns.IsDynamicNode(ecfg.Address) && ecfg.Group {
+			return EndpointConfig{}, errors.Newf("%s is a dynamically specified endpoint i.e. it uses SD and that is not permitted under group mode.", ecfg.Address)
+		}
+	}
+
+	if !slices.Contains([]string{snappy.Name, compressionNone}, cfg.Compression) {
+		return EndpointConfig{}, errors.Newf("compression must be one of [%q, %q], got: %s", snappy.Name, compressionNone, cfg.Compression)
+	}
+
+	return EndpointConfig{}, nil
+}
+
+func newEndpointReloader(
+	g *run.Group,
+	logger log.Logger,
+	configFile *extflag.PathOrContent,
+	configReloadInterval time.Duration,
+) (*endpointConfigReloader, error) {
+	res := &endpointConfigReloader{}
+
+	cfg, err := res.parse(configFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load config file")
+	}
+	res.cfg = cfg
+
+	reloadCtx, reloadCancel := context.WithCancel(context.Background())
+	g.Add(func() error {
+		return extkingpin.PathContentReloader(reloadCtx, configFile, logger, func() {
+			res.mu.Lock()
+			defer res.mu.Unlock()
+
+			level.Info(logger).Log("msg", "reloading endpoint config")
+			cfg, err := res.parse(configFile)
+			if err != nil {
+				level.Error(logger).Log("msg", "failed to reload endpoint config", "err", err)
+				return
+			}
+			res.cfg = cfg
+		}, configReloadInterval)
+	}, func(err error) {
+		reloadCancel()
+	})
+
+	return res, nil
+}
+
+type configProvider interface {
+	config() EndpointConfig
+}
+
+type providerFunc func() EndpointConfig
+
+func (f providerFunc) config() EndpointConfig { return f() }
+
+// For querier
+func setupEndpointset(
+	g *run.Group,
+	reg prometheus.Registerer,
+	logger log.Logger,
+	configFile *extflag.PathOrContent,
+	configReloadInterval time.Duration,
+	dnsSDResolver string,
+	dnsSDInterval time.Duration,
+	unhealthyTimeout time.Duration,
+	endpointTimeout time.Duration,
+	queryConnMetricLabels ...string,
+) (*query.EndpointSet, error) {
+	reloader, err := newEndpointReloader(g, logger, configFile, configReloadInterval)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load config initially")
+	}
+	return setupEndpointsetWithProvider(
+		g,
+		reg,
+		logger,
+		reloader,
+		dnsSDResolver,
+		dnsSDInterval,
+		unhealthyTimeout,
+		endpointTimeout,
+		queryConnMetricLabels...,
+	)
+}
+
+// For ruler ( for some reason )
```

Comment:
What do you mean? :smile: Ruler supports querying Query through gRPC. That's why it is needed

Reply:
Yeah but its configured way differently; essentially it only gets list of addresses not all the other endpoint config flags

Reply:
I added this comment because in ideal world ruler should just use same config like querier here

Reply:
Can we clarify this comment so that it would be clear for future readers?

Reply:
Yeah! I think we could even add this "--endpoint.sd-config" flag to ruler too eventually in future PR and unify a bit with the   querier

Reply:
Added a lengthy comment, opposed to the former "for some reason"!